File: SWEA/D4/3124. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Solution {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    static int[] parents;
    static int V, E;
    static Edge[] edgeList;

    static class Edge implements Comparable<Edge> {
        int start, end, weight;

        public Edge(int start, int end, int weight) {
            super();
            this.start = start;
            this.end = end;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge o) {
            return Integer.compare(this.weight, o.weight);
        }
    }

    public static void main(String[] args) throws IOException {
        int t = Integer.parseInt(br.readLine());

        for (int tc = 1; tc <= t; tc++) {
            st = new StringTokenizer(br.readLine(), " ");
            V = Integer.parseInt(st.nextToken());
            E = Integer.parseInt(st.nextToken());

            edgeList = new Edge[E];

            for (int i = 0; i < E; i++) {
                st = new StringTokenizer(br.readLine(), " ");
                int start = Integer.parseInt(st.nextToken()) - 1;
                int end = Integer.parseInt(st.nextToken()) - 1;
                int weight = Integer.parseInt(st.nextToken());
                edgeList[i] = new Edge(start, end, weight);
            }
            Arrays.sort(edgeList);

            init();

            int cnt = 0;
            long result = 0;
            for (Edge edge : edgeList) {
                if (union(edge.start, edge.end)) {
                    result += edge.weight;
                    if (++cnt == V - 1) break;
                }
            }
            bw.write(String.format("#%d %d\n", tc, result));
        }
        bw.close();
    }

    private static boolean union(int a, int b) {
        int aRoot = find(a);
        int bRoot = find(b);
        if (aRoot == bRoot) return false;

        parents[bRoot] = aRoot;
        return true;
    }

    private static int find(int a) {
        if (a == parents[a]) return a;

        return parents[a] = find(parents[a]);
    }

    private static void init() {
        parents = new int[V];

        for (int i = 0; i < V; i++) {
            parents[i] = i;
        }
    }
}

File: SWEA/D4/3124. 최소 스패닝 트리/최소 스패닝 트리.py
import collections


def find(x):
    if x == parent[x]:
        return x
    parent[x] = find(parent[x])

    return parent[x]


def union(a, b):
    a = find(a)
    b = find(b)

    if a == b:
        return False

    if rank[a] > rank[b]:
        parent[b] = a
    elif rank[a] < rank[b]:
        parent[a] = b
    else:
        parent[b] = a
        rank[a] += 1

    return True


for t in range(1, int(input()) + 1):
    V, E = map(int, input().split())

    graph = []

    for _ in range(E):
        a, b, c = map(int, input().split())
        graph.append((a, b, c))

    graph.sort(key=lambda x: x[-1])
    parent = collections.defaultdict(int)
    rank = collections.defaultdict(int)

    for i in range(1, V + 1):
        parent[i] = i

    result = 0
    for a, b, c in graph:
        if union(a, b):
            result += c

    print(f'#{t} {result}')
